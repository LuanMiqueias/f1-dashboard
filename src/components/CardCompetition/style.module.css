
.card_competition{
  width: 100%;
  max-width: 250px;
  transition: all .2s;
  box-shadow: 0px 3px 2px rgba(0, 0, 0, 0.15);

}
.card_competition:hover{
  transition: all .07s;
  transform: translateY(-5px);
  box-shadow: 0px 3px 4px 1px rgba(0, 0, 0, 0.2);

}
.content_card_competition{
  width: 100%;
  height: 70px;
  display: flex;
  align-items: flex-start;
  justify-content: center;
  flex-direction: column;
  gap:4px;
  padding: 1rem 2rem;
  border-radius: 5px;
  cursor: pointer;
  background: var(--white);
  position: relative;
  user-select: none;
  border:1px solid rgb(238, 238, 238);
  border-bottom: 3px solid var(--primary-color);
}

.content_card_competition h2{
  font-weight: 400;
  font-size: 1rem;
  color:var(--sub-title);
  max-width: 200px;
  text-overflow: ellipsis;
  white-space: nowrap;
  overflow: hidden;
}
.content_card_competition p{
  font-weight: 400;
  font-size: .75rem;
  color:var(--text-light);
  max-width: 200px;
  text-overflow: ellipsis;
  white-space: nowrap;
  overflow: hidden;
}