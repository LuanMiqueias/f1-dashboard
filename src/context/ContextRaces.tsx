import React, { ReactNode } from "react";
import api from "../services/api";

interface PropsRaces {
  id: number;
  distance?: string;
  date: string;
  status: string;
  season: number;
  type: string;
  laps: {
    total?: number;
  };
  circuit: {
    id: number;
    name: string;
    image: string;
  };
}
interface IData {
  data: PropsRaces[];
}
interface RacesContent {
  races: IData;
  idCompetition: number;
  oneRace: PropsRaces;
  loading: boolean;
  error: string;
  selectRace: (id: number) => void;
  updateRaces: (data: IData) => void;
  updateIdCompetition: (id: number) => void;
}
export const RacesContext = React.createContext({} as RacesContent);

export const RacesProvider: React.FC<ReactNode> = ({ children }) => {
  const [races, setRaces] = React.useState(null as IData);
  const [idCompetition, setIdCompetition] = React.useState(null as number);
  const [oneRace, setOneRace] = React.useState(null as PropsRaces);
  const [loading, setLoading] = React.useState(true as boolean);
  const [error, setError] = React.useState("" as string);

  React.useEffect(() => {
    //DeverÃ¡ set feito um Fetch para api usando o ID da competicao
    if (idCompetition) {
      setLoading(true);
      (async () => {
        const response = await api.get(
          `races?competition=${idCompetition}&season=2021`,
          {
            headers: {
              "x-rapidapi-host": "v1.formula-1.api-sports.io",
              "x-rapidapi-key": process.env.NEXT_PUBLIC_API_KEY,
            },
          }
        );
        if (response.data.errors.request) {
          setError(
            "Sorry for the inconvenience, this site uses a free API plan and you or someone else has exceeded the limit ðŸ˜­, please try again later."
          );
        }
        setRaces({ data: response.data.response });
        setLoading(false);
      })();
    }
  }, [idCompetition]);

  function updateRaces(data: IData) {
    setRaces(data);
    setOneRace(null);
  }
  function updateIdCompetition(id: number) {
    setIdCompetition(id);
    setOneRace(null);
  }
  function selectRace(id: number) {
    if (races) {
      const race = races.data.filter((item) => {
        return item.id === id;
      });
      setOneRace(race[0]);
    }
  }
  return (
    <RacesContext.Provider
      value={{
        races,
        idCompetition,
        oneRace,
        loading,
        error,
        selectRace,
        updateRaces,
        updateIdCompetition,
      }}
    >
      {children}
    </RacesContext.Provider>
  );
};
