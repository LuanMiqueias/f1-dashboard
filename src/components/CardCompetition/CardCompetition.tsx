import React from "react";
import { RacesContext } from "../../context/ContextRaces";
import Card from "../Card";
import styles from "./style.module.css";

interface PropsDrives {
  id: number;
  name: string;
  location: {
    country: string;
    city: string;
  };
}

const CardCompetition = ({ id, name, location }: PropsDrives) => {
  const { updateIdCompetition } = React.useContext(RacesContext);

  return (
    <div
      className={styles.card_competition}
      onClick={() => updateIdCompetition(id)}
    >
      <div className={styles.content_card_competition}>
        <h2>{name}</h2>
        <p>
          {location.country}, {location.city}
        </p>
      </div>
    </div>
  );
};

export default CardCompetition;
