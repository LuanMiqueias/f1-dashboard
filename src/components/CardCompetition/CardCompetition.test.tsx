import { fireEvent, render, screen } from "@testing-library/react";
import React from "react";
import { RacesContext, RacesProvider } from "../../context/ContextRaces";
import CardCompetition from "./CardCompetition";
const fakeData = {
  id: 2070,
  name: "name1",
  location: {
    country: "country1",
    city: "city1",
  },
};
it("Should have snapshot", () => {
  const { container } = render(<CardCompetition {...fakeData} />);
  expect(container).toMatchSnapshot();
});

it("Shoud have props(name, country, city)", () => {
  render(<CardCompetition {...fakeData} />);
  expect(screen.getByRole("heading", { name: /name1/i })).toBeInTheDocument;
  expect(screen.getByText(/city1/i)).toBeInTheDocument;
  expect(screen.getByText(/country1/i)).toBeInTheDocument;
});

it("Shoud called function and set idCompetition", async () => {
  const contextCallback = jest.fn();
  const { container } = render(
    <RacesProvider>
      <CardCompetition {...fakeData} />
      <RacesContext.Consumer>{contextCallback}</RacesContext.Consumer>
    </RacesProvider>
  );
  fireEvent.click(container.firstElementChild);
  expect(contextCallback.mock.calls[1][0].idCompetition).toEqual(2070);
});
