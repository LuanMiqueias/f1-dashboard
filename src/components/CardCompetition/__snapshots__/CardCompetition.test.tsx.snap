// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should have snapshot 1`] = `
<div>
  <div
    class="card_competition"
  >
    <div
      class="content_card_competition"
    >
      <h2>
        name1
      </h2>
      <p>
        country1
        , 
        city1
      </p>
    </div>
  </div>
</div>
`;
